Web Inspector: Debugger: add a way to step over `await` as though it was sync code

dcrousso opened the pull request:
#### 6eceba5825819069e719f9b9a560ca6c1a64b3e8
<pre>
Web Inspector: Debugger: add a way to step over `await` as though it was sync code
<a href="https://bugs.webkit.org/show_bug.cgi?id=291309">https://bugs.webkit.org/show_bug.cgi?id=291309</a>

Reviewed by Yusuke Suzuki.

When debugging async code with `await` it&apos;s often more desirable to follow what&apos;s written in code (i.e. as if the `await` didnt exist) instead of the literal execution flow, especially since it&apos;s possible for execution to stop after the `await` if it&apos;s the end of sync code.

For example, if the developer is trying to step through code similar to the following
```js
/* 1 */ async function wrap(callback, state) {
/* 2 */     console.log(state, &quot;before&quot;);
/* 3 */     await callback();
/* 4 */     console.log(state, &quot;after&quot;);
/* 5 */ }
/* 6 */
/* 7 */ wrap(foo, 1);
/* 8 */ debugger;
/* 9 */ wrap(foo, 2);
```
execution should never pause in the flow where `state === 1`, meaning that the debugger will &quot;skip&quot; (i.e. resume and then pause again) between &quot;2 before&quot; and &quot;1 after&quot;.

Without this, the developer has to manually create breakpoints after each `await` to ensure that execution pauses after, but even then there&apos;s no guarantee that will be in the right flow if the same function is invoked multiple times (i.e. one resolves before the other).

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_debug):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::operationDebug):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::slow_path_debug):
Modify `OpDebug` to allow for additional data to be set alongside the `DebugHookType`.
Currently this is only used to pass along the `JSGenerator` instance as a way to link execution from before the `await` to after the `await`.
Drive-by: Remove the unused `hasBreakpoint` member variable.

* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::debug):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.cpp:
(JSC::dumpExpressionInfoDetails):
* Source/JavaScriptCore/debugger/Debugger.h:
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::detach):
(JSC::Debugger::pauseIfNeeded):
(JSC::Debugger::handlePause):
(JSC::Debugger::willAwait): Added.
(JSC::Debugger::didAwait): Added.
(JSC::Debugger::resetAsyncPauseState): Added.
Introduce new `WillAwait` and `DidAwait` debug hooks for the `Debugger` to determine what `await` to pause on.
When performing a &quot;Step over&quot; or &quot;Step next&quot;, if an `await` is encountered then save the `JSGenerator` instance (which is the additional data set alongside the `DebugHookType`).
Once the `await` is done, if the `JSGenerator` matches what was saved then pause at the next opportunity.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitDebugHook):
(JSC::BytecodeGenerator::emitWillLeaveCallFrameDebugHook):
(JSC::BytecodeGenerator::emitAwait):
(JSC::BytecodeGenerator::emitIteratorGenericNext):
(JSC::BytecodeGenerator::emitIteratorGenericClose):
(JSC::BytecodeGenerator::emitDelegateYield):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ReturnNode::emitBytecode):
(JSC::emitProgramNodeBytecode):
(JSC::EvalNode::emitBytecode):
(JSC::FunctionNode::emitBytecode):
(JSC::AwaitExprNode::emitBytecode):
Drive-by: remove unnecessary overrides to make debug hook logic simpler.

* Source/WebKitLegacy/mac/WebView/WebScriptDebugger.h:
* Source/WebKitLegacy/mac/WebView/WebScriptDebugger.mm:
(WebScriptDebugger::handlePause):
Drive-by: no need to pass the `ReasonForPause` since it&apos;s already exposed via `Debugger::reasonForPause`.

* LayoutTests/inspector/debugger/resources/stepping-async.js: Added.
* LayoutTests/inspector/debugger/stepping/stepInto-await.html:
* LayoutTests/inspector/debugger/stepping/stepInto-await-expected.txt:
* LayoutTests/inspector/debugger/stepping/stepNext-await.html:
* LayoutTests/inspector/debugger/stepping/stepNext-await-expected.txt:
* LayoutTests/inspector/debugger/stepping/stepOut-await.html:
* LayoutTests/inspector/debugger/stepping/stepOut-await-expected.txt:
* LayoutTests/inspector/debugger/stepping/stepOver-await.html:
* LayoutTests/inspector/debugger/stepping/stepOver-await-expected.txt:

Canonical link: <a href="https://commits.webkit.org/293628@main">https://commits.webkit.org/293628@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ba8d6354147665451801f1ef6580610e134397d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99434 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19084 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104565 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50035 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19372 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75683 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32779 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102441 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14739 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89781 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56042 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14534 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7764 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49395 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92117 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84455 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106923 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98053 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26548 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19366 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84642 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26910 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85985 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84159 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21352 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28830 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6521 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20291 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26488 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31689 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121669 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26308 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33982 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29621 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27875 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4974,8 +4969,12 @@ RegisterID* BytecodeGenerator::emitYield(RegisterID* argument)
-RegisterID* BytecodeGenerator::emitAwait(RegisterID* dst, RegisterID* src)
+RegisterID* BytecodeGenerator::emitAwait(RegisterID* dst, RegisterID* src, const JSTextPosition& position)
 {
+    static unsigned lastAwaitIdentifier = 0;
```
i have a suspicion that this wont work, but not entirely sure what scenarios would be problematic 😅

@Constellation any ideas?

Constellation replied:
I think this is problematic since we are assuming that same code will produce the same bytecode. As a result, if we have compiled this twice (once for non debugger and once with debugger), we will have a bad time. At least, this needs to be deterministic per function.

dcrousso replied:
this value needs to be globally unique in order for the debugger to be able to differentiate between two calls to the same function

for example, if we have the following code
```js
/* 1 */ async function wrap(callback, state) {
/* 2 */     console.log(state, "before");
/* 3 */     await callback();
/* 4 */     console.log(state, "after");
/* 5 */ }
/* 6 */ 
/* 7 */ debugger;
/* 8 */ wrap(foo, 1);
/* 9 */ wrap(foo, 2);
```
and the developer does a "Step into" on line 8 and then repeatedly does a "Step over", execution should never pause in the flow where `state === 2`

tho since this is generated in the bytecode, would that even give a unique identifier? 🤔

Constellation commented on Source/JavaScriptCore/bytecode/BytecodeList.rb:
```diff
@@ -1133,7 +1133,7 @@
     args: {
         debugHookType: DebugHookType,
-        hasBreakpoint: bool,
+        data?: unsigned,
```
Do we need this `?` ? I think it is not necessary.

dcrousso replied:
is this not a way to represent "this field is optional"?  it's fine to always provide it, but i was thinking that it didnt have to always exist

burg commented on Source/JavaScriptCore/debugger/Debugger.cpp:
```diff
@@ -1210,6 +1234,45 @@ void Debugger::atExpression(CallFrame* callFrame)
+    if (m_isPaused)
+        return;
+
+    if (!m_pauseAfterAwaitFromGenerator || m_pauseAfterAwaitFromGenerator != generator)
```
Since This name is kind of ambiguous without more context- I would prefer m_generatorBeingSteppedThrough or similar. As written, I am led to believe it could refer to a boolean (should it pause after await?) or a Generator that we store for awaiting.

burg commented on Source/JavaScriptCore/debugger/Debugger.cpp:
```diff
@@ -1210,6 +1234,45 @@ void Debugger::atExpression(CallFrame* callFrame)
+    if (!m_pauseAfterAwaitFromGenerator || m_pauseAfterAwaitFromGenerator != generator)
+        return;
+
+    m_pauseAfterAwaitFromGenerator = 0;
```
🤔

burg reviewed (commented):
I took a first look at this and don't have any major concerns other than that raised already. I will take another look at the tests.

Constellation commented on Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
```diff
@@ -4019,31 +4019,26 @@ void BytecodeGenerator::emitDebugHook(DebugHookType debugHookType, const JSTextP
 
     emitExpressionInfo(divot, divot, divot);
-    OpDebug::emit(this, debugHookType, false);
+    OpDebug::emit(this, debugHookType, data ? VirtualRegister(data) : VirtualRegister());
```
Empty `VirtualRegister` is not allowed. I recommend setting jsUndefined constant register in this case, like this.
Also we do not need to wrap RegisterID* with VirtualRegister.
```
if (!data)
    data = emitLoad(nullptr, jsUndefined());
OpDebug::emit(this, debugHookType, data);
```

Constellation commented on Source/JavaScriptCore/interpreter/Interpreter.h:
```diff
@@ -157,7 +159,7 @@ using JSOrWasmInstruction = std::variant<const JSInstruction*, const WasmInstruc
         NEVER_INLINE CatchInfo unwind(VM&, CallFrame*&, Exception*);
         void notifyDebuggerOfExceptionToBeThrown(VM&, JSGlobalObject*, CallFrame*, Exception*);
-        NEVER_INLINE void debug(CallFrame*, DebugHookType);
+        NEVER_INLINE void debug(CallFrame*, DebugHookType, JSCell* data);
```
Let's just take JSValue for now as we are passing jsUndefined when it is not a generator.

Constellation commented on Source/JavaScriptCore/jit/JITOpcodes.cpp:
```diff
@@ -1668,7 +1668,8 @@ void JIT::emit_op_debug(const JSInstruction* currentInstruction)
     load32(Address(regT0, CodeBlock::offsetOfDebuggerRequests()), regT0);
     Jump noDebuggerRequests = branchTest32(Zero, regT0);
-    callOperation(operationDebug, TrustedImmPtr(&vm()), static_cast<int>(bytecode.m_debugHookType));
+    emitGetVirtualRegisterPayload(bytecode.m_data, regT1);
```
Let's just get the data via `emitGetVirtualRegister`. And storing it to `jsRegT32`.

Constellation commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -2889,14 +2889,14 @@ JSC_DEFINE_JIT_OPERATION(operationHandleTraps, UnusedPtr, (JSGlobalObject* globa
 }
 
-JSC_DEFINE_JIT_OPERATION(operationDebug, void, (VM* vmPointer, int32_t debugHookType))
+JSC_DEFINE_JIT_OPERATION(operationDebug, void, (VM* vmPointer, int32_t debugHookType, JSCell* data))
```
And taking it as `EncodedJSValue encodedData`.

Constellation commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -2889,14 +2889,14 @@ JSC_DEFINE_JIT_OPERATION(operationHandleTraps, UnusedPtr, (JSGlobalObject* globa
     auto scope = DECLARE_THROW_SCOPE(vm);
 
-    vm.interpreter.debug(callFrame, static_cast<DebugHookType>(debugHookType));
+    vm.interpreter.debug(callFrame, static_cast<DebugHookType>(debugHookType), data);
```
Pass it via `JSValue::decode(encodedData)`.

Constellation commented on Source/JavaScriptCore/debugger/Debugger.h:
```diff
@@ -345,6 +331,7 @@ class Debugger : public DoublyLinkedListNode<Debugger> {
     JSValue m_currentException;
     CallFrame* m_pauseOnCallFrame { nullptr };
     CallFrame* m_currentCallFrame { nullptr };
+    JSGenerator* m_pauseInGenerator { nullptr };
```
How is this object kept alive from GC?

dcrousso replied:
fundamentally all we're using this for is as a pointer identifier (i.e. we never dereference it)

generally speaking we also really don't like it when Web Inspector keeps things alive as that can make it much more difficult for developers to identify memory issues

would wrapping it in `Weak` work?

dcrousso committed:
Web Inspector: Debugger: add a way to step over `await` as though it was sync code
https://bugs.webkit.org/show_bug.cgi?id=291309

Reviewed by Yusuke Suzuki.

When debugging async code with `await` it's often more desirable to follow what's written in code (i.e. as if the `await` didnt exist) instead of the literal execution flow, especially since it's possible for execution to stop after the `await` if it's the end of sync code.

For example, if the developer is trying to step through code similar to the following
```js
/* 1 */ async function wrap(callback, state) {
/* 2 */     console.log(state, "before");
/* 3 */     await callback();
/* 4 */     console.log(state, "after");
/* 5 */ }
/* 6 */
/* 7 */ wrap(foo, 1);
/* 8 */ debugger;
/* 9 */ wrap(foo, 2);
```
execution should never pause in the flow where `state === 1`, meaning that the debugger will "skip" (i.e. resume and then pause again) between "2 before" and "1 after".

Without this, the developer has to manually create breakpoints after each `await` to ensure that execution pauses after, but even then there's no guarantee that will be in the right flow if the same function is invoked multiple times (i.e. one resolves before the other).

* Source/JavaScriptCore/bytecode/BytecodeList.rb:
* Source/JavaScriptCore/bytecode/BytecodeUseDef.cpp:
(JSC::computeUsesForBytecodeIndexImpl):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::emit_op_debug):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::operationDebug):
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::slow_path_debug):
Modify `OpDebug` to allow for additional data to be set alongside the `DebugHookType`.
Currently this is only used to pass along the `JSGenerator` instance as a way to link execution from before the `await` to after the `await`.
Drive-by: Remove the unused `hasBreakpoint` member variable.

* Source/JavaScriptCore/interpreter/Interpreter.h:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::debug):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/UnlinkedCodeBlock.cpp:
(JSC::dumpExpressionInfoDetails):
* Source/JavaScriptCore/debugger/Debugger.h:
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::detach):
(JSC::Debugger::pauseIfNeeded):
(JSC::Debugger::handlePause):
(JSC::Debugger::willAwait): Added.
(JSC::Debugger::didAwait): Added.
(JSC::Debugger::resetAsyncPauseState): Added.
Introduce new `WillAwait` and `DidAwait` debug hooks for the `Debugger` to determine what `await` to pause on.
When performing a "Step over" or "Step next", if an `await` is encountered then save the `JSGenerator` instance (which is the additional data set alongside the `DebugHookType`).
Once the `await` is done, if the `JSGenerator` matches what was saved then pause at the next opportunity.

* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.h:
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::emitDebugHook):
(JSC::BytecodeGenerator::emitWillLeaveCallFrameDebugHook):
(JSC::BytecodeGenerator::emitAwait):
(JSC::BytecodeGenerator::emitIteratorGenericNext):
(JSC::BytecodeGenerator::emitIteratorGenericClose):
(JSC::BytecodeGenerator::emitDelegateYield):
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::ReturnNode::emitBytecode):
(JSC::emitProgramNodeBytecode):
(JSC::EvalNode::emitBytecode):
(JSC::FunctionNode::emitBytecode):
(JSC::AwaitExprNode::emitBytecode):
Drive-by: remove unnecessary overrides to make debug hook logic simpler.

* Source/WebKitLegacy/mac/WebView/WebScriptDebugger.h:
* Source/WebKitLegacy/mac/WebView/WebScriptDebugger.mm:
(WebScriptDebugger::handlePause):
Drive-by: no need to pass the `ReasonForPause` since it's already exposed via `Debugger::reasonForPause`.

* LayoutTests/inspector/debugger/resources/stepping-async.js: Added.
* LayoutTests/inspector/debugger/stepping/stepInto-await.html:
* LayoutTests/inspector/debugger/stepping/stepInto-await-expected.txt:
* LayoutTests/inspector/debugger/stepping/stepNext-await.html:
* LayoutTests/inspector/debugger/stepping/stepNext-await-expected.txt:
* LayoutTests/inspector/debugger/stepping/stepOut-await.html:
* LayoutTests/inspector/debugger/stepping/stepOut-await-expected.txt:
* LayoutTests/inspector/debugger/stepping/stepOver-await.html:
* LayoutTests/inspector/debugger/stepping/stepOver-await-expected.txt:

Canonical link: https://commits.webkit.org/293628@main
